File: webtool/drust_client.py
import time

import requests

# Observed live: Python's `requests` occasionally hits a transient
# connection failure against the Drust host even when curl succeeds
# immediately after (same machine, same network). A single blip shouldn't
# fail a 30-50 minute translation request outright.
_MAX_ATTEMPTS = 3
_RETRY_BACKOFF_SECONDS = 1


class DrustClient:
    def __init__(self, base_url: str, tenant_id: str, anon_token: str, service_token: str):
        self.base_url = base_url.rstrip("/")
        self.tenant_id = tenant_id
        self.anon_token = anon_token
        self.service_token = service_token

    def _tenant_path(self, path: str) -> str:
        return f"{self.base_url}/t/{self.tenant_id}/{path}"

    def _post_with_retry(self, url: str, json_body: dict, headers: dict) -> requests.Response:
        last_error = None
        for attempt in range(_MAX_ATTEMPTS):
            try:
                return requests.post(url, json=json_body, headers=headers, timeout=30)
            except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as e:
                last_error = e
                if attempt < _MAX_ATTEMPTS - 1:
                    time.sleep(_RETRY_BACKOFF_SECONDS)
        raise last_error

    def fetch_glossary(self) -> list[dict]:
        rows = []
        page = 1
        per_page = 200
        total = None
        while True:
            resp = self._post_with_retry(
                self._tenant_path("collections/translation_glossary/list"),
                {"page": page, "per_page": per_page},
                {"Authorization": f"Bearer {self.anon_token}"},
            )
            resp.raise_for_status()
            data = resp.json()
            batch = data.get("records", [])
            if not batch:
                break
            rows.extend(batch)
            if total is None:
                total = data.get("total")
            if total is not None and len(rows) >= total:
                break
            page += 1
        return rows

    def insert_pending_term(self, term: str, stage: str, context: str,
                             suggested_fix: str, video_title: str) -> dict:
        resp = self._post_with_retry(
            self._tenant_path("records/pending_terms"),
            {
                "data": {
                    "term": term,
                    "stage": stage,
                    "context": context,
                    "suggested_fix": suggested_fix,
                    "video_title": video_title,
                },
            },
            {"Authorization": f"Bearer {self.service_token}"},
        )
        resp.raise_for_status()
        data = resp.json()
        # Live tenant wraps the created row under "record"; fall back to the
        # top-level payload for tenants/mocks that return the row directly.
        return data.get("record", data)

File: cf-worker/src/drustClient.ts
import { GlossaryRow } from "./promptBuilder";

const MAX_ATTEMPTS = 3;
const RETRY_BACKOFF_MS = 1000;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class DrustClient {
  private baseUrl: string;
  private tenantId: string;
  private anonToken: string;
  private serviceToken: string;

  constructor(baseUrl: string, tenantId: string, anonToken: string, serviceToken: string) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.tenantId = tenantId;
    this.anonToken = anonToken;
    this.serviceToken = serviceToken;
  }

  private tenantPath(path: string): string {
    return `${this.baseUrl}/t/${this.tenantId}/${path}`;
  }

  private async postWithRetry(url: string, jsonBody: unknown, headers: Record<string, string>): Promise<Response> {
    let lastError: unknown;
    for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
      try {
        return await fetch(url, {
          method: "POST",
          headers: { ...headers, "Content-Type": "application/json" },
          body: JSON.stringify(jsonBody),
        });
      } catch (e) {
        lastError = e;
        if (attempt < MAX_ATTEMPTS - 1) {
          await sleep(RETRY_BACKOFF_MS);
        }
      }
    }
    throw lastError;
  }

  async fetchGlossary(): Promise<GlossaryRow[]> {
    const rows: GlossaryRow[] = [];
    let page = 1;
    const perPage = 200;
    let total: number | null = null;

    while (true) {
      const resp = await this.postWithRetry(
        this.tenantPath("collections/translation_glossary/list"),
        { page, per_page: perPage },
        { Authorization: `Bearer ${this.anonToken}` }
      );
      if (!resp.ok) {
        throw new Error(`Drust list failed: ${resp.status} ${await resp.text()}`);
      }
      const data = (await resp.json()) as { records?: GlossaryRow[]; total?: number };
      const batch = data.records ?? [];
      if (batch.length === 0) break;
      rows.push(...batch);
      if (total === null) total = data.total ?? null;
      if (total !== null && rows.length >= total) break;
      page += 1;
    }
    return rows;
  }

  async insertPendingTerm(params: {
    term: string;
    stage: string;
    context: string;
    suggestedFix: string;
    videoTitle: string;
  }): Promise<Record<string, unknown>> {
    const resp = await this.postWithRetry(
      this.tenantPath("records/pending_terms"),
      {
        data: {
          term: params.term,
          stage: params.stage,
          context: params.context,
          suggested_fix: params.suggestedFix,
          video_title: params.videoTitle,
        },
      },
      { Authorization: `Bearer ${this.serviceToken}` }
    );
    if (!resp.ok) {
      throw new Error(`Drust insert failed: ${resp.status} ${await resp.text()}`);
    }
    const data = (await resp.json()) as { record?: Record<string, unknown> };
    return data.record ?? (data as Record<string, unknown>);
  }
}
